src/javascript/_common/autocomplete_address.js

```diff
@@ -0,0 +1,184 @@
+                    marker.setPosition(results[0].geometry.location);
+
+                    autocomplete = new google.maps.places.Autocomplete(
+                        /** @type {!HTMLInputElement} */(el_address_line_1),
+                        {
+                            types                : ['geocode'],
+                            componentRestrictions: {country: residence},
+                        });
+
+                    // Bind the map's bounds (viewport) property to the autocomplete object,
+                    // so that the autocomplete requests use the current map bounds for the
+                    // bounds option in the request.
+                    autocomplete.bindTo('bounds', map);
+
+                    autocomplete.addListener('place_changed', () => {
+                        const place = autocomplete.getPlace();
+                        fillInAddress(place);
+                    });
+
+                    if (el_address_line_1.value) {
+                        geocoder.geocode({ address: el_address_line_1.value}, (res, stat) => {
+                            if (stat === google.maps.GeocoderStatus.OK) {
+                                fillInAddress(res[0]);
+                            }
+                        });
+                    }
+                }
+            });
+        });
+
+        const onDragEnd = () => {
+            geocoder.geocode({ location: marker.getPosition() }, (results, status) => {
+                if (status === google.maps.GeocoderStatus.OK) {
+                    fillInAddress(results[0]);
+                }
+            });
+        };
+    };
+
+    /*
+     * Fill in address form
+     *
+     * @param {Object} place - google autocomplete object
+     */
+    const fillInAddress = (place) => {
+        const el_address_line_1   = document.getElementById('address_line_1');
+        const el_address_line_2   = document.getElementById('address_line_2');
+        const el_address_city     = document.getElementById('address_city');
+        const el_address_state    = document.getElementById('address_state');
+        const el_address_postcode = document.getElementById('address_postcode');
+
+        const address_line_1_data = [];
+        const address_line_2_data = [];
+        let has_city = false;
+
+        // Get each component of the address from the place details
+        // and fill the corresponding field on the form.
+        for (let i = 0; i < place.address_components.length; i++) {
+            const address_type = place.address_components[i].types[0];
+            if (component_form[address_type]) {
+                const val = place.address_components[i][component_form[address_type]];
+                if (/street_number|route|neighborhood/.test(address_type)) {
+                    address_line_1_data.push(val);
+                }
+                if (/(administrative_area_level_3|administrative_area_level_4|administrative_area_level_5)/.test(address_type)) {
```

Comment:
`/(administrative_area_level_3|administrative_area_level_4|administrative_area_level_5)/` can be shortened to `/administrative_area_level_(3|4|5)/`